### ISSUE TITLE
Flow doesn't recognize switch-break statements within conditional blocks

### ISSUE THREAD

**Issue by buoyad** [facebook/flow#6068]
Example:
```
// @flow

type A = 'yellow' | 'orange'

type holdsA = {
  type: A
}

const processA = (h: holdsA) => {
  switch (h.type) {
      case 'yellow':
          if (Math.round(Math.random()) === 1) {
        	break
          }
      case 'orange':
          if (Math.round(Math.random()) === 1) {
        	break
          }
      default:
          throw new Error('Bad!')
  }
  console.log('we made it')
}	
```

When run on this snippet ([playground](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AQVALygDkaW0c8xoAPibAE4CGAdgOZbHLragAWsaABMAzviIBvRAEgeWAFx5EAX2QBjWCxEpQGBrDVYRYwqAAUfRQOFiAlIQB8oKaFAj4ASxRq+5vgDo5exdXULUmERxSckpieURQxNCPSHMAWSYUAP0AVxYhMwys-2Z82ABbM1t7AlrQAEZghKTXaQAjBiwmAGtmltdVFvDI+lKOOL7+lPTM7Ng8gqLs1iEKqpq6xudJxPbOnp2kwaShLEgmHOgUeP7QrP14UBYsR4BRBn0GM2IAISYhACExFsk2OGi0giw-jgbG+8Bw5X+OC8wJU0kQQA)), flow gives the following error:

```
    22:   console.log('we made it')
          ^ Unreachable code.
```

Although there is a definite chance that `Math.round(Math.random()) === 1` will be true and the switch will be broken out of. Indeed, running the function in console reaches the log statement sometimes:

<img width="311" alt="image" src="https://user-images.githubusercontent.com/11968340/38145812-2bcb3e7a-3419-11e8-9a48-c01592e23be7.png">

Moving any of the `break`s to the top level of the `case` statement gets rid of the error, as well as removing the `throw` in the default case. This leads me to think that flow isn't recognizing the `break`s within the conditionals as valid execution paths, and is assuming that the default case will be triggered every time.

I think this is a common enough pattern for people manually doing type refinements within switch statements, and is pretty annoying to work around without using `$FlowIssue`s (and even then it's bad, because flow marks every line after the end of the switch statement). Regardless, flow should recognize that the switch statement can be broken out of and should not be marking the code following it as unreachable.

**Comment by thehogfather** [comment]
I think what flow is trying to tell you in this case is that you are missing your else clauses in each case statement. Don't forget flow (and probably most type systems) have no way of understanding that `Math.round(Math.random()) === 1` will resolve to `true`. All it knows is that it resolves to a boolean which produces two branches.

**Comment by buoyad** [comment]
@thehogfather good point about flow not knowing about whether that condition will be true, however, [changing the predicate](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AQVALygDkaW0c8xoAPibAE4CGAdgOZbHLragAWsaABMAzviIBvRKBmgeWAFx5EAX2QBjWCxEpQGBrHVYRYwqAAUfJQOFiAlIQB8oKTJHwAlinV8LfAHTyDq6yMupMIjik5JTECtKhoR6QFigMAK5YwQmJiQBGDFhMANY5uWq5QliQTOnQKPG5oSh8BvCgLFjtAKIMBgzmxABCTEIAhMR2CRWgmtqCWP5wbIPwOAC2ozhek6qgQA) doesn't solve the problem:

```
// @flow

type A = 'yellow' | 'orange'

type holdsA = {
    type: A
}

const processA = (h: holdsA) => {
  switch (h.type) {
      case 'yellow':
         if (true) {
            break
         }
      default:
          throw new Error('Bad!')
  }
  console.log('we made it')
} 
```

Flow still gives `unreachable code` here when the predicate is explicitly true. 

Regardless of the predicate though, `Unreachable code` is code that cannot be executed "because there exists no control flow path to the code from the rest of the program" [[1]](https://en.wikipedia.org/wiki/Unreachable_code). This definition means that flow doesn't need to understand anything about which control flow is going to run (though it'd be nice if it recognized that substituting `if (false)` above would cause a problem); as long as a branch exists that leads to a piece of code, that code shouldn't be labeled as unreachable.

**Comment by thehogfather** [comment]
@buoyad as far as I understand, in this case the type system sees a conditional statement where all cases have not been explicitly handled (ignore for a second that you are explicitly setting it to true, false or some predicate whose value you know). If you had an else (from where you also break), then this would be fine. Admittedly the error message should probably be more like `Possibly unreachable code` because if the predicate in your `if` branch were false, then it falls through to the default case and the function would `throw`.  

If you don't mind me asking, if the condition is always true, why wrap it in an if?

**Comment by buoyad** [comment]
@thehogfather the actual conditional isn't always true; that (^) is just a convenient example to show the absurdity of the error. Other than that, I'm not sure we're talking about the same thing. Of course it's _possible_ that the `console.log` won't be reached, but `Unreachable code` is code that by definition can _never_ run, regardless of what conditionals evaluate to.

```
const cb = () => {
  return
  console.log('we made it') // <-- truly unreachable code
}
```

This issue is about flow labeling the (original example) code after the switch statement as `Unreachable` when that is plainly false - the code can be reached!

**Comment by gkz** [comment]
We no longer have the error "  ^ Unreachable code." for the `console.log`: https://flow.org/try/#1N4Igxg9gdgZglgcxALlAIwIZoKYBsD6uEEAztvhgE6UYCe+JADpdhgCYowa5kA0I2KAFcAtiRQAXSkOz9sADwxgJ+NPTbYuQ3BMnTZA+Y2yU4IwRO4A6SFBIrGVDGM7c+h46fNRLuKxJIGWh8MeT0ZfhYlCStpHzNsFBAMIQkIEQwJODAQfiEyfBE4eWw2fDgofDBMsAALfAA3KjgsXGxxZC4eAw0G-GhcWn9aY3wWZldu-g1mbGqJUoBaCRHEzrcDEgBrbAk62kXhXFxJ923d-cPRHEpTgyEoMDaqZdW7vKgoOfaSKgOKpqmDA+d4gB5fMA-P6LCCMLLQbiLOoYCqgh6-GDYRYIXYLSgkRZkCR4jpddwPfJLZjpOBkO4AX34kA0SQAOj5VgACACCnIAvJyAOS0PBEADugs5AB8hRAaFAcYL2eyVsZObUILg2CReQLgOzOZzVdhkDz2fTlY9oPZOdTISQdfzOQAKWqmjVanUASn5AD5OfqoIaSGK4Htai7asNjD7A4b45zqmQhSLjhAJcgDQns3AYC6ALKZKOUCAPNjOwsSYvAtjpZ1en18pucgCMsaz2c5rIkaCiWw7nc5FqD2aT2Fl8sVmZHg85uYLRdipag5cr1ZXdYb-ObbYDA4T3d7rH7M8Hw87Gi0Omns8NVZLYs5X0fAFFqHLnYKAELsACEgq9LNz0Ta1NWwKwiAQT8xXHDINDnCQAPNbtchABoTBIOBoCSBoAAYrAAJgANgAZisXCQHpIA
